tutor/hooks/catalog.py

```diff
@@ -258,6 +258,26 @@ def your_filter_callback(some_data):
     #:   arguments. Note that these arguments do not include the leading ``docker`` command.
     DOCKER_BUILD_COMMAND: Filter[list[str], []] = Filter()
 
+    #: List of python package names that can be potentially installed next to edx-platform.
+    #: Whenever a user runs: ``tutor mounts add /path/to/name`` "name" will be matched to
+    #: the regular expressions in this filter. If it matches, then the directory will be
+    #: automatically bind-mounted in the "openedx" Docker image at build time and run
+    #: time. They will be mounted in ``/openedx/requirements/<name>``. Then, ``pip install
+    #: -e .`` will be run in this directory at build-time. And the same host directory
+    #: will be bind-mounted in that location at run time. This allows users to
+    #: transparently work on edx-platform dependencies.
+    #:
+    #: By default, xblocks and some common packages are already present in this
+    #: filter. Add your own edx-platform dependencies to this filter to make it easier for
+    #: users to work on edx-platform dependencies.
+    #:
+    #: See the list of all edx-platform base requirements here:
+    #: https://github.com/openedx/edx-platform/blob/master/requirements/edx/base.txt
+    #:
+    #: :parameter list[str] names: Add here simple names ("my-package") or regular
+    #:   expressions.
+    EDX_PLATFORM_PYTHON_PACKAGES: Filter[list[str], []] = Filter()
+
```

Comment:
Two things:

1. This filter hook is openedx-specific, unlike every other hook in catalog.py. Are you sure you want to add this hook here, given that you are extracting openedx-specific code into its own module in this very same PR?

2. As devs migrate over from devstack, they will need to hack on packages for services other than edx-platform. I guess we could adopt a pattern where each service plugin defines its own version of this filter (e.g., `DISCOVERY_PYTHON_PACKAGES: ... = Filter()`). Alternatively, we could make this filter's name more generic, e.g. `KNOWN_PYTHON_PACKAGES: Filter[list[str], []]`, and each plugin could decide how to process the filter. This would also ensure that catalog.py doesn't have openedx-specific stuff in it.

Reply:
> As devs migrate over from devstack, they will need to hack on packages for services other than edx-platform. 

As an example, I know @rgraber has been needing to run course-discovery with a local Python package in order to work on the event bus, and had found the old Tutor workflow difficult.

Reply:
A couple notes for myself:

1. Moving the filter to a separate module would make it difficult to document it. Also, developers would have to import a separate module, e.g: `tutor.openedx.hooks`.
2. We could change the filter signature to `Filter[list[tuple[str, str]], []]` to make it slightly more generic. But it would still be openedx-specific, so we should still move it to a separate module.

I don't have a great answer for now. I'll think about it some more.

Reply:
I may have found a good solution. It yields the following documentation:

![image](https://github.com/overhangio/tutor/assets/44319/83cf761b-e280-4d7e-8f9e-90c27bd7fde6)

Let me know what you think.

Reply:
@regisb I'm not sure that solves the right problems. Tutor's docs already have plenty of Open-edX-specific details, and that's OK (probably good, actually. Otherwise, the docs would be confusingly abstract).

The issues I see with EDX_PLATFORM_PYTHON_PACKAGES is that (1) it makes it harder for us to isolate an Open-edX-agnostic core of Tutor's _codebase_, and (2) it prevents micro-services like Credentials and Course Discovery from using this improvement.

What about this?
```python
    #: List of python package names that can be potentially installed into a particular image.
    #: ...
    #: :parameter list[str] names: Add here simple names ("my-package") or regular
    #:   expressions.
    #: :parameter [str] image: Name of the image (e.g., "openedx") which should install
    #:   the packages, if mounted.
    IMAGES_PYTHON_PACKAGES: Filter[list[str], [str]] = Filter()
```

One wrinkle is that package names would need to be added to this filter for both prod and dev images (e.g., openedx and openedx-dev). We could, of course, write a low-priority filter which adds every package from openedx to openedx-dev.

Reply:
> The issues I see with EDX_PLATFORM_PYTHON_PACKAGES is that (1) it makes it harder for us to isolate an Open-edX-agnostic core of Tutor's codebase, and (2) it prevents micro-services like Credentials and Course Discovery from using this improvement.

Right, I agree. Which is why I moved the filter to a dedicated openedx module in my latest changes. I wasn't so keen on doing that because I figured it would make it more difficult to document the new hooks, but the solution is acceptable (I think?).

Would you prefer I keep the hook in the tutor.hooks.catalog module?

Reply:
Oooook, I had misunderstood you. Your proposition makes sense, I'll update my implementation.

Reply:
I have yet again updated my implementation. It feels quite generic and it can be used by any plugin developer -- though it's definitely very advanced usage.